### ISSUE TITLE
cloud-init fails on Amazon Linux and CentOS AMI's, "Error persisting instance-data.json"

### ISSUE THREAD

**Issue by aaroniscode** [kubernetes-sigs/cluster-api-provider-aws#983]
/kind bug

**What steps did you take and what happened:**
Create a CAPA cluster and choose an Amazon Linux AMI. I tried `ami-092d9d6a36468257d` which is for k8s v1.14.4.

I also tried with a custom built AMI for k8s 1.15.2 and both fail with the same error.

cloud-init fails very early in the process in `sources/__init__.py`.
```2019-08-08 12:31:38,951 - __init__.py[WARNING]: Error persisting instance-data.json: 'utf8' codec can't decode byte 0x8b in position 1: invalid start byte```

Because `instance-data-json` is not written, cloud-init has no user-data and the PKI assets aren't written and `kubeadm` isn't run.

**What did you expect to happen:**
cloud-init to write PKI assets and run `kubeadm init`

**Anything else you would like to add:**
Spent a good deal of time troubleshooting and trying to determine root cause. Here is what I learned.
The contents of the user-data are fine. I logged into the instance, downloaded the gzip user-data and extracted it. Stopped the ec2 instance, updated the user-data by pasting the user-data as text into the EC2 console. Restarted the node and forced cloud-init to run again. PKI assets were created and kubeadm was run.

I compared the non-working Amazon Linux node to a working Ubuntu node. Ubuntu's cloud-init is based on Python3. Amazon Linux's cloud-init is based on Python 2. cloud-init versions are very close. I don't remember the exact versions.

Similar issue on Azure with cloud-init failing to persist instance-data.json: https://bugzilla.redhat.com/show_bug.cgi?id=1631167

Guesses on root cause (as far as I got at this point)
* Maybe the user-data has a non-ascii character in the dictionary
* Perhaps Python3 handles this situation better than Python2

**Environment:**

- Cluster-api-provider-aws version: v0.3.7
- Kubernetes version: (use `kubectl version`): N/A, doesn't seem dependent on k8s version
- OS (e.g. from `/etc/os-release`): Amazon Linux 2

**Comment by randomvariable** [comment]
There is some string parsing differences between Python versions so very well could be related to that. We would expect CentOS 7 to break in the same way.

**Comment by mauilion** [comment]
I think @randomvariable is on the right track:

Further evidence: https://bugs.launchpad.net/cloud-init/+bug/1801364

A work around for this may be to bring the version of python installed up to the "known working" version.

**Comment by aaroniscode** [comment]
Tested on CentOS and it fails with the same error as Amazon Linux 2.
```2019-08-08 19:05:50,543 - __init__.py[WARNING]: Error persisting instance-data.json: 'utf8' codec can't decode byte 0x8b in position 1: invalid start byte```

This was with `ami-0cf28ce2e4d1c46bf` from https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/docs/amis.md#centos-7-1

**Comment by aaroniscode** [comment]
This issue might be related: https://github.com/hashicorp/terraform/issues/4794#issuecomment-176858792

**Comment by randomvariable** [comment]
Thanks for investigating. I'll keep track of support
/assign

**Comment by randomvariable** [comment]
/remove-lifecycle active
/close

We intend to fix this as part of v1alpha2, so closing in favour of https://github.com/kubernetes-sigs/image-builder/issues/17
